chain/network/src/network_protocol/proto_conv/peer_message.rs

```diff
@@ -569,3 +594,70 @@ impl TryFrom<&proto::PeerMessage> for PeerMessage {
         })
     }
 }
+
+impl From<&PeerIdOrHash> for proto::PeerIdOrHash {
+    fn from(x: &PeerIdOrHash) -> Self {
+        match x {
+            PeerIdOrHash::PeerId(peer_id) => proto::PeerIdOrHash {
+                target_type: Some(proto::peer_id_or_hash::Target_type::PeerId(peer_id.into())),
+                ..Default::default()
+            },
+            PeerIdOrHash::Hash(hash) => proto::PeerIdOrHash {
+                target_type: Some(proto::peer_id_or_hash::Target_type::Hash(hash.into())),
+                ..Default::default()
+            },
+        }
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ParsePeerIdOrHashError {
+    #[error("missing target_type")]
+    MissingTargetType,
+    #[error("peer_id: {0}")]
+    PeerId(ParsePublicKeyError),
+    #[error("hash: {0}")]
+    Hash(ParseCryptoHashError),
+}
+
+impl TryFrom<&proto::PeerIdOrHash> for PeerIdOrHash {
+    type Error = ParsePeerIdOrHashError;
+    fn try_from(x: &proto::PeerIdOrHash) -> Result<Self, Self::Error> {
+        match x.target_type.as_ref().ok_or(Self::Error::MissingTargetType)? {
+            proto::peer_id_or_hash::Target_type::PeerId(peer_id) => {
```

Comment:
nit: maybe import `proto::peer_id_or_hash::Target_type::*` to reduce boilerplate